packages/SwingSet/src/liveslots/virtualReferences.js

```diff
@@ -160,6 +165,46 @@ export function makeVirtualReferenceManager(
     kindInfoTable.set(`${kindID}`, { reanimator, deleter, durable });
   }
 
+  /**
+   * Compare two arrays (shallowly) for equality.
+   *
+   * @template T
+   * @param {T[]} a1
+   * @param {T[]} a2
+   * @returns {boolean}
+   */
+  function arrayEquals(a1, a2) {
+    assert(Array.isArray(a1));
+    assert(Array.isArray(a2));
+    if (a1.length !== a2.length) {
+      return false;
+    }
+    return a1.every((elem, idx) => a2[idx] === elem);
+  }
+
+  /**
+   * Check a list of facet names against what's already been established for a
+   * kind.  If they don't match, it's an error.  If nothing has been established
+   * yet, establish it now.
```

Comment:
Do you already have something like this for the field names from an `init`, to ensure they are uniform across a kind? Or do you allow these to differ per instance?

Reply:
That's actually an interesting question.  We really do have to enforce the invariance of the set of facets so that the object's API can't be mutated into different shapes depending on when it's deserialized.  However, the virtual object manager allows different instances to be initialized with different field names, and since any given instance is only initialized once it'll work fine so long as the relevant methods are expecting a variant structure.  However, we've never contemplated making any use of this "feature" and it's arguably a bad coding practice.  Enforcing field name uniformity might be a good idea (certainly it would be easy to do), but we haven't made a decision and we should.

Reply:
I'd like to at least announce/document that Kinds should produce the same facets (and the same state property names) on all instances, so if/when we manage to enforce it later, users will have been warned.

Fixed facet/state names gives us more to work with, and we could imagine converting that data to SQL table schema.

Reply:
I could have sworn that enforced requirement that facets be invariant was documented, but I just looked and it wasn't (it is now, or anyway will be with the next commit).  The open question is whether we should enforce invariance on properties as well.  (It also strikes me that we should ask the same question about the methods of the facets, but then we'd be getting into wether the method functions were the same functions each time, which is basically impossible to check).